crates/telio-network-monitors/src/monitor.rs

```diff
@@ -0,0 +1,127 @@
+                            Ok(v) => {
+                                if let Ok(mut guard) = LOCAL_ADDRS_CACHE.lock() {
+                                    telio_log_debug!("Updating local addr cache");
+                                    *guard = v;
+                                }
+                            }
+                            Err(e) => telio_log_warn!("Unable to get local interfaces {e}"),
+                        },
+                        Err(e) => {
+                            telio_log_warn!("Failed to receive new path for sockets ({e}): ");
+                        }
+                    }
+                }
+            }
+        }));
+
+        #[cfg(any(target_os = "macos", target_os = "ios", target_os = "tvos"))]
+        NETWORK_PATH_MONITOR_START.call_once(crate::mac::setup_network_monitor);
+
+        #[cfg(target_os = "linux")]
+        let linux_monitor_handle = crate::linux::setup_network_monitor().await;
+
+        #[cfg(target_os = "windows")]
+        let windows_monitor_handle = crate::windows::setup_network_monitor();
+
+        Ok(Self {
+            nw_path_monitor_monitor_handle,
+            #[cfg(target_os = "linux")]
+            linux_monitor_handle,
+            #[cfg(target_os = "windows")]
+            windows_monitor_handle,
+        })
+    }
+}
+
+impl Drop for NetworkMonitor {
+    fn drop(&mut self) {
+        if let Some(handle) = &self.nw_path_monitor_monitor_handle {
+            handle.abort();
+        }
+        #[cfg(target_os = "linux")]
+        if let Some(handle) = &self.linux_monitor_handle {
+            handle.abort();
+        }
+        #[cfg(target_os = "windows")]
+        if !self.windows_monitor_handle.0.is_null() {
+            crate::windows::deregister_network_monitor(self.windows_monitor_handle.clone());
+        }
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use telio_utils::local_interfaces::MockGetIfAddrs;
+
+    use super::*;
+    use std::net::Ipv4Addr;
+
+    #[tokio::test]
+    async fn test_path_change_notification() {
+        let mut get_if_addrs_mock = MockGetIfAddrs::new();
+        get_if_addrs_mock.expect_get().times(2).returning(|| {
+            Ok(vec![if_addrs::Interface {
+                name: "old".to_owned(),
+                addr: if_addrs::IfAddr::V4(if_addrs::Ifv4Addr {
+                    ip: Ipv4Addr::new(10, 0, 0, 1),
+                    netmask: Ipv4Addr::new(255, 255, 255, 0),
+                    broadcast: None,
+                }),
+                index: None,
+                #[cfg(windows)]
+                adapter_name: "{78f73923-a518-4936-ba87-2a30427b1f63}".to_string(),
+            }])
+        });
+        let network_monitor = NetworkMonitor::new(get_if_addrs_mock).await.unwrap();
+
+        if let Err(e) = PATH_CHANGE_BROADCAST.send(()) {
+            println!("Failed to notify about changed path, error: {e}");
+        }
+        tokio::time::sleep(tokio::time::Duration::from_secs(1)).await;
```

Comment:
why do we need to wait at the end of test?

Reply:
We need as it takes a bit of time for the callback/interrupt to occur from the OS and then monitor notifying it since the purpose of this test is to mainly see whether send() triggers the if.get().